### PR TITLE
Bench 256

### PR DESCRIPTION
**Pull request by martintaylor1635**
- Aligned Categories with the contract
- Refactored Tests

### CODE REVIEW

File: src/main/java/com/answerdigital/answerking/service/CategoryService.java [hunk 1]
```diff
@@ -36,84 +37,173 @@ public CategoryService(final ProductService productService,
...
-            throw new NameUnavailableException(String.format("A category named '%s' already exists", categoryRequest.name()));
-        }
+        validateCategoryNameDoesNotExist(categoryRequest.name(), Optional.empty());
 
-        final Category newCategory = categoryMapper.addRequestToCategory(categoryRequest);
-        final var category =  categoryRepository.save(newCategory);
-        return categoryMapper.convertCategoryEntityToCategoryResponse(category);
+        // Create a new category from the Request and persist the initial category.
+        final Category newCategory = requestToCategory(categoryRequest);
+        final Category category =  categoryRepository.save(newCategory);
```
Comment (AndyFlintAnswerDigital) [inline src/main/java/com/answerdigital/answerking/service/CategoryService.java:52]: I'd simplify this by just having it be 

`Category category = requestToCategory(categoryRequest);
category =  categoryRepository.save(newCategory);`

File: src/main/java/com/answerdigital/answerking/service/CategoryService.java [hunk 2]
```diff
@@ -36,84 +37,173 @@ public CategoryService(final ProductService productService,
...
+        }
+    }
+
+    /**
+     * Checks if a Category already exists with a given name.
+     * @param categoryName The Category name.
+     * @param id The ID of the Category.
+     * @throws NameUnavailableException When the Category name already exists.
+     */
+    private void validateCategoryNameDoesNotExist(final String categoryName, final Optional<Long> id) {
```
Comment (AndyFlintAnswerDigital) [inline src/main/java/com/answerdigital/answerking/service/CategoryService.java:170]: Nested ifs almost always sit wrong with me. I'd argue that we should have this as two separate methods. One for when we're renaming categories and one for when we're creating new ones.

**Review by AndyFlintAnswerDigital — CHANGES_REQUESTED** [review]
I've not been able to finish the entire review (namely the 3 large classes) but have added a few comments.

File: src/main/java/com/answerdigital/answerking/controller/CategoryController.java [hunk 3]
```diff
@@ -88,28 +88,6 @@ public ResponseEntity<Collection<ProductResponse>> fetchProductsByCategory(@Path
...
-    @ApiResponses(value = {
-        @ApiResponse(responseCode = "200", description = "Remove product from a category.",
-                content = { @Content(mediaType = "application/json", schema = @Schema(implementation = CategoryResponse.class)) })
-    })
-    @PutMapping("/{categoryId}/removeproduct/{productId}")
-    public ResponseEntity<CategoryResponse> removeProductFromCategory(@PathVariable @NotNull final Long categoryId,
-                                                              @PathVariable @NotNull final Long productId) {
-        return new ResponseEntity<>(categoryService.removeProductFromCategory(categoryId, productId), HttpStatus.OK);
-    }
-
```
[the remark this answers is not in the document]
Comment (martintaylor1635) [reply src/main/java/com/answerdigital/answerking/controller/CategoryController.java:112]: Resolved

File: src/main/java/com/answerdigital/answerking/service/CategoryService.java (hunk 1 above)
Comment (martintaylor1635) [reply src/main/java/com/answerdigital/answerking/service/CategoryService.java]: Resolved

File: src/main/java/com/answerdigital/answerking/service/CategoryService.java (hunk 2 above)
Comment (martintaylor1635) [reply src/main/java/com/answerdigital/answerking/service/CategoryService.java]: Resolved

**Review by oleksiiRybak — APPROVED** [review]

**Pull request merged**